File: js/services/income-service.js
import { database } from '../firebase-config.js';
import { ref, push, set, get, remove } from 'firebase/database';

export class IncomeService {
    constructor() {
        this.path = 'income';
    }

    async create(income) {
        const incomeRef = ref(database, this.path);
        const newRef = push(incomeRef);
        await set(newRef, income);
        return newRef.key;
    }

    async update(id, income) {
        const incomeRef = ref(database, `${this.path}/${id}`);
        await set(incomeRef, income);
    }

    async delete(id) {
        const incomeRef = ref(database, `${this.path}/${id}`);
        await remove(incomeRef);
    }

    async get(id) {
        const incomeRef = ref(database, `${this.path}/${id}`);
        const snapshot = await get(incomeRef);
        if (snapshot.exists()) {
            return { id: snapshot.key, ...snapshot.val() };
        }
        return null;
    }

    async getAll() {
        const incomeRef = ref(database, this.path);
        const snapshot = await get(incomeRef);
        const incomes = [];
        if (snapshot.exists()) {
            snapshot.forEach(child => {
                incomes.push({ id: child.key, ...child.val() });
            });
        }
        return incomes;
    }

    async getByDateRange(startDate, endDate) {
        const incomes = await this.getAll();
        return incomes.filter(income => {
            const incomeDate = new Date(income.date);
            return incomeDate >= startDate && incomeDate <= endDate;
        });
    }

    async getTotal(startDate, endDate) {
        const incomes = await this.getByDateRange(startDate, endDate);
        return incomes.reduce((sum, income) => sum + (income.amount || 0), 0);
    }
}

export const incomeService = new IncomeService();

File: lib/services/income_service.dart
import 'dart:async';
import '../models/income.dart';
import 'firebase_service.dart';

class IncomeService {
  final FirebaseService _firebase;
  final String _path = 'income';

  IncomeService(this._firebase);

  Future<String> create(Income income) async {
    return await _firebase.save(_path, income.toJson());
  }

  Future<void> update(Income income) async {
    if (income.id == null) throw Exception('Income ID is required');
    await _firebase.update(_path, income.id!, income.toJson());
  }

  Future<void> delete(String id) async {
    await _firebase.delete(_path, id);
  }

  Future<Income?> get(String id) async {
    final data = await _firebase.get(_path, id);
    return data != null ? Income.fromJson(data, id) : null;
  }

  Future<List<Income>> getAll() async {
    final data = await _firebase.getAll(_path);
    return data.map((item) => Income.fromJson(item, item['id'])).toList();
  }

  Future<List<Income>> getByDateRange(DateTime start, DateTime end) async {
    final incomes = await getAll();
    return incomes.where((income) {
      return income.date.isAfter(start.subtract(const Duration(days: 1))) &&
          income.date.isBefore(end.add(const Duration(days: 1)));
    }).toList();
  }

  Future<double> getTotal(DateTime start, DateTime end) async {
    final incomes = await getByDateRange(start, end);
    return incomes.fold<double>(0.0, (double sum, Income income) => sum + income.amount);
  }
}


